toolchain/check/deduce.cpp

```diff
@@ -0,0 +1,196 @@
+        worklist.deductions.pop_back_val();
+
+    // If the parameter has a symbolic type, deduce against that.
+    auto param_type_id = context.insts().Get(param_id).type_id();
+    if (param_type_id.AsConstantId().is_symbolic()) {
+      worklist.Add(
+          context.types().GetInstId(param_type_id),
+          context.types().GetInstId(context.insts().Get(arg_id).type_id()),
+          needs_substitution);
+    }
+
+    // If the parameter is a symbolic constant, deduce against it.
+    auto param_const_id = context.constant_values().Get(param_id);
+    if (!param_const_id.is_valid() || !param_const_id.is_symbolic()) {
+      continue;
+    }
+
+    // If we've not yet substituted into the parameter, do so now.
+    if (needs_substitution) {
+      param_const_id = SubstConstant(context, param_const_id, substitutions);
+      if (!param_const_id.is_valid() || !param_const_id.is_symbolic()) {
+        continue;
+      }
+      needs_substitution = false;
+    }
+
+    CARBON_KIND_SWITCH(context.insts().Get(context.constant_values().GetInstId(
+                           param_const_id))) {
+      // Deducing a symbolic binding from an argument with a constant value
+      // deduces the binding as having that constant value.
+      case CARBON_KIND(SemIR::BindSymbolicName bind): {
+        auto& entity_name = context.entity_names().Get(bind.entity_name_id);
+        auto index = entity_name.bind_index;
+        if (index.is_valid() && index >= first_deduced_index) {
+          CARBON_CHECK(static_cast<size_t>(index.index) < results.size())
+              << "Deduced value for unexpected index " << index
+              << "; expected to deduce " << results.size() << " arguments.";
+          auto arg_const_inst_id =
+              context.constant_values().GetConstantInstId(arg_id);
+          if (arg_const_inst_id.is_valid()) {
+            if (results[index.index].is_valid() &&
+                results[index.index] != arg_const_inst_id) {
+              // TODO: Include the two different deduced values.
+              CARBON_DIAGNOSTIC(DeductionInconsistent, Error,
+                                "Inconsistent deductions for value of generic "
+                                "parameter `{0}`.",
+                                SemIR::NameId);
+              auto diag = context.emitter().Build(
+                  node_id, DeductionInconsistent, entity_name.name_id);
+              NoteGenericHere(context, generic_id, diag);
+              diag.Emit();
+              return SemIR::SpecificId::Invalid;
+            }
+            results[index.index] = arg_const_inst_id;
+          }
+        }
+        break;
+      }
+
+        // TODO: Handle more cases.
+
+      default:
+        break;
+    }
+  }
+
+  // Check we deduced an argument value for every parameter.
+  for (auto [i, deduced_arg_id] : llvm::enumerate(results)) {
+    if (!deduced_arg_id.is_valid()) {
+      auto binding_id = context.inst_blocks().Get(
```

Comment:
```suggestion
      CARBON_CHECK(i >= first_deduced_index);
      auto binding_id = context.inst_blocks().Get(
```

This diagnostic seems like it would be confusing if it referred to a parameter of the enclosing generic, and presumably that's not supposed to happen anyway.

Reply:
Yeah, good point. Changed to only check the deduced arguments.